src/i2c.rs

```diff
@@ -0,0 +1,268 @@
+        &self.buffer
+    }
+}
+
+impl DerefMut for I2cSlaveWriteBuffer {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.buffer
+    }
+}
+
+pub struct I2cDriver<'a> {
+    lifetime: PhantomData<&'a ()>,
+}
+
+impl<'a> I2cDriver<'a> {
+    pub fn init_master_write_buffer(
+        &self,
+        buffer: &'a mut I2cMasterWriteBuffer,
+    ) -> TockResult<SharedMemory> {
+        syscalls::allow(DRIVER_NUMBER, allow_nr::MASTER_WRIE, &mut buffer.buffer)
+            .map_err(Into::into)
+    }
+
+    pub fn init_master_read_buffer(
+        &self,
+        buffer: &'a mut I2cMasterReadBuffer,
+    ) -> TockResult<SharedMemory> {
+        syscalls::allow(DRIVER_NUMBER, allow_nr::MASTER_READ, &mut buffer.buffer)
+            .map_err(Into::into)
+    }
+
+    pub fn init_slave_read_buffer(
+        &self,
+        buffer: &'a mut I2cSlaveReadBuffer,
+    ) -> TockResult<SharedMemory> {
+        syscalls::allow(DRIVER_NUMBER, allow_nr::SLAVE_READ, &mut buffer.buffer).map_err(Into::into)
+    }
+
+    pub fn init_slave_write_buffer(
+        &self,
+        buffer: &'a mut I2cSlaveWriteBuffer,
+    ) -> TockResult<SharedMemory> {
+        syscalls::allow(DRIVER_NUMBER, allow_nr::SLAVE_WRITE, &mut buffer.buffer)
+            .map_err(Into::into)
+    }
+
+    pub fn subscribe<CB: FnMut(usize, usize)>(
+        &self,
+        callback: &'a mut CB,
+    ) -> TockResult<CallbackSubscription> {
+        syscalls::subscribe::<I2cEventConsumer, _>(
+            DRIVER_NUMBER,
+            subscribe_nr::SUBSCRIBE_CALLBACK,
+            callback,
+        )
+        .map_err(Into::into)
+    }
+
+    pub fn check_present(&self) -> TockResult<()> {
+        syscalls::command(DRIVER_NUMBER, command_nr::CHECK_PRESENT, 0, 0)?;
+        Ok(())
+    }
+
+    pub fn master_write(&self, address: usize, length: usize) -> TockResult<()> {
+        syscalls::command(
+            DRIVER_NUMBER,
+            command_nr::MASTER_WRITE,
+            address & 0xFFFF | length << 16,
+            0,
+        )?;
+        Ok(())
+    }
+
+    pub fn master_read(&self, address: usize, length: usize) -> TockResult<()> {
+        syscalls::command(
+            DRIVER_NUMBER,
+            command_nr::MASTER_READ,
+            address & 0xFFFF | length << 16,
+            0,
+        )?;
+        Ok(())
+    }
+
+    pub fn slave_listen(&self) -> TockResult<()> {
+        syscalls::command(DRIVER_NUMBER, command_nr::SLAVE_LISTEN, 0, 0)?;
+        Ok(())
+    }
+
+    pub fn slave_enable_read(&self) -> TockResult<()> {
+        syscalls::command(DRIVER_NUMBER, command_nr::SLAVE_ENABLE_READ, 0, 0)?;
+        Ok(())
+    }
+
+    pub fn slave_stop_listen(&self) -> TockResult<()> {
+        syscalls::command(DRIVER_NUMBER, command_nr::SLAVE_STOP_LISTEN, 0, 0)?;
+        Ok(())
+    }
+
+    pub fn slave_set_address(&self, address: usize) -> TockResult<()> {
+        syscalls::command(DRIVER_NUMBER, command_nr::SLAVE_SET_ADDRESS, address, 0)?;
+        Ok(())
```

Comment:
`check_present`, `master_write`, `master_read`, ..., `slave_set_address` look like they should be 1-liners, not 2-liners (i.e. that they should return the output of `command` directly rather than using `?;` and `Ok(())`).